(DOCSP-26779): Rules v2 (#324)
mongodb/docs-app-services: source/rules/roles.txt

How App Services Assigns Roles
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

App Services assigns roles at different times depending on whether you're using

> :ref:`Device Sync (Flexible Mode) <sync>` or not.

Reviewer: This section only deals with Flexible Sync, correct? The Partition-Based Sync stuff is covered elsewhere.

I think we could be clearer here by replacing most instances of "Device Sync" with "Flexible Sync". This makes it clearer that this doesn't work for Partition-Based Sync.

Author: It's tricky because I'm trying to subtly make "Device Sync" mean "Flexible Sync", as it should eventually supersede PBS.

Reviewer: Ah, I see. I think we run the risk of confusing some people, but the number is probably pretty small. Might be worth it to gradually shift Device Sync perception the direction you're going.

For each document returned by the query, your App evaluates possible
roles in :ref:`role order <role-order>` and assigns the first applicable
role, if any. A role applies to a given document if its "apply when"
:ref:`expression <expressions>` evaluates to ``true`` when run against
the document.

The set of roles that a user can use for a given request depends on the
collection they're accessing. If the collection has a set of

> collection-level roles defined, then a user may play any of those roles.

Reviewer: I'm curious about the wording change here. "use" doesn't seem accurate, but "play" seems odd too. I've typically seen users and roles described in a couple of ways:
- the user is the role: a user may **be** any of those roles
- the user has a role: a user may **have** any of those roles

Author: Yes! I inherited this verbiage, and I fixed it, and it seems my fix was undone in a merge. I will re-apply.

              ]
            }

With Device Sync
~~~~~~~~~~~~~~~~

> When using Device Sync, App Services assigns a role at the beginning of each
> Sync session. The role determines which permissions apply for the duration of a
> Flexible Sync session.

Reviewer: Specifying Flexible Sync in the second sentence, but not in the first feels odd.
```suggestion
When using Device Sync, App Services assigns a role at the beginning of each
Flexible Sync session. The role determines which permissions apply for the duration of a
session.
```

example, if the user's metadata or the role's "apply when" expression changes,
the user continues to use their existing session role until the next time they
start a session.

There are some special considerations when using Device Sync with regard to the
permissions system. See :ref:`device-sync-compatibility`.

For a guide to setting up Flexible Sync with common permissions models, see the
:ref:`flexible-sync-permissions-guide`.

.. _apply-when-expressions:

Apply When Expressions
~~~~~~~~~~~~~~~~~~~~~~

A role's "apply when" expression is a :ref:`rule expression <expressions>` that
determines whether the role should be assigned.

You can use expression variables to make roles dynamic. For example, you can use the
:json-expansion:`%%user` expansion to refer to the specific user that issued the
request. This lets you customize your data access permissions on a per-user
basis.

When not using Device Sync, you can refer to the current document for which a
role is being assigned. For example, you can use :json-expansion:`%%root`. This
lets you customize your data access permissions on a per-document basis.

.. _document-filters-expressions:

Document Filters
~~~~~~~~~~~~~~~~

> The role's ``document_filters`` expressions determine whether the role's

Reviewer 2: (feel free to resolve this thread)

roles remain.

A user may only have one role per document in a given query. Role order
determines which role applies in the event that multiple roles' "apply when"
expressions are true. Therefore, when defining roles, put the most specific
roles first.

Sync Compatibility
~~~~~~~~~~~~~~~~~~

If Device Sync (Flexible Mode) is enabled, an assigned role must be **sync
compatible**. See :ref:`sync-compatible-roles` for details.

.. _define-roles-and-permissions:
.. _define-roles:

Define Roles & Permissions
--------------------------

You can configure your app's data access rules from the App Services UI or by
deploying configuration files with {+cli+}:

.. tabs-realm-admin-interfaces::

   .. tab::
      :tabid: ui

      1. Click :guilabel:`Rules` in the left navigation menu and then select a
         collection from the data source's collection list. You can also select
         :guilabel:`Default roles and filters` to configure default rules for
         the entire cluster.
      
         If no roles are defined yet, you'll be prompted to create a new
         one. Otherwise you'll see an ordered list of existing roles.
         

>       #. To create a new role, click :guilabel:`Add role`. You can
>          either start from scratch or use a preset role as a starting
>          point.

Reviewer: Let's keep these steps parallel by removing the infinitive. Creating a new role is the implied goal of this procedure. If it's not clear enough, we should add a procedure title above all the steps.

We also probably want to avoid the "start from scratch" idiom, as it may be harder for non-native English speakers to understand.

```suggestion
      #. Click :guilabel:`Add role` to define a new role or or use a 
         preset role as a starting point.
```

         
      #. Give the role a name. The name can be anything you want but
         must be unique within a given collection. Consider using names

>          that describe the user (e.g. ``Admin``) and/or their
>          relationship to the data (e.g. ``Owner``).

Reviewer: It's typically a good idea to avoid Latin terms and abbreviations to improve readability.
```suggestion
         that describe the user and/or their relationship to the data. For 
         example, ``Admin`` or ``Owner``.
```

      
      #. Define an "apply when" expression that determines when a given user has
         the role for a given document.
      

>       #. Define :ref:`document filters <document-filters-expressions>` for the

Reviewer 2: (feel free to resolve this thread)

         role.

      #. Define :ref:`document-level permissions <document-level-permissions>`
         for the role.
      
      #. Define :ref:`field-level permissions <field-level-permissions>` for the
         role.
      
      #. Save the role.

      #. If a collection has more than one role assigned, you can modify

>          their :ref:`role order <role-order>` by clicking the arrows on

Reviewer: Slight semantic change to make the connection to collection clearer. Otherwise, it may seem like the roles themselves have a role order.
```suggestion
         the :ref:`role order <role-order>` by clicking the arrows on
```
